### DISCUSSION TITLE
How to pass a composite type into a postgresql function?

### DISCUSSION THREAD

**Discussion by wwuck** [sqlalchemy/sqlalchemy#10095] [Usage Questions] [answered]
I'm trying to work out how to pass either a composite type or array of composite types into a postgres function using sqlalchemy 2.0 ORM.

Previously the codebase I'm working on was using [`CompositeType`](https://sqlalchemy-utils.readthedocs.io/en/latest/data_types.html#module-sqlalchemy_utils.types.pg_composite) from `sqlalchemy-utils` with sqlalchemy 1.3/1.4 but I'm trying to convert it to use SQLAlchemy 2.0 ORM.  This is mostly due to an attempt to switch from `psycopg2` engine to `pg8000` engine which isn't supported by `sqlalchemy-utils CompositeType`.

Are there any examples available on how to do this?

EDIT: 
using postgresql composite types https://www.postgresql.org/docs/current/rowtypes.html
example:
```sql
CREATE TYPE c_data AS (c_id bigint,date date);
```
```sql
CREATE OR REPLACE FUNCTION build_c(
    c_data_row c_data[]
)
RETURNS TABLE(
    f_c_id bigint
    ,f_date date
)
LANGUAGE plpgsql

COST 100
VOLATILE
ROWS 1000
AS $BODY$

DECLARE cd public.c_data;

BEGIN
    FOREACH cd IN ARRAY c_data_row LOOP
        -- ...
    END LOOP;
END;   
        
$BODY$;
```

**Comment by CaselIT** [comment]
Hi,

Composite are a python only abstraction, so they don't exist on the db.
Can you clarify what you mean by "pass a composite type to a postgresql function"?

**Reply by wwuck** [reply]
Sorry for the confusion, I've added an example of the postgresql function.  Is this unrelated to the `sqlalchemy.orm.composite()`?

**Reply by CaselIT** [reply]
Yes, that's a way of mapping multiple columns inside a python class. On the db the storage of the single column is still separate

**Reply by wwuck** [reply]
Hmm ok thanks. So it’s still only supported by sqlalchemy-utils?

**Reply by CaselIT** [reply]
I think so, yes.

**Reply by wwuck** [reply]
That’s unfortunate because sqlalchemy-utils is hardcoded to only support psycopg2. No other engines are supported, not even psycopg3.

**Reply by zzzeek** [reply]
I would report to sqlalchemy-utils that they should modernize and support more backends

**Reply by wwuck** [reply]
https://github.com/kvesteri/sqlalchemy-utils/issues/712
https://github.com/kvesteri/sqlalchemy-utils/issues/575

**Reply by andrew222651** [reply]
Would be nice functionality for sqlalchemy to have, just being able to call a function, maybe open an issue

**Reply by CaselIT** [reply]
Proposals for this could be considered.
There is support for other named types like domain and enums. Will add pr welcome label

**Reply by andrew222651** [reply]
In raw SQL you can write a value with a composite type as `CAST(ROW(field1, field2) AS CompositeTypeName)`. In SQLAlchemy, I guess you can do `func.row(*fields)` but how can you cast to an arbitrary `CompositeTypeName`? Does `execute(text("cast(:row as CompositeTypeName)"), {"row": func.row(*fields)})` work?

ETA: this works, at least in asyncpg, although the arguments have to be "literal" objects:
```python
func.funcname(text("cast(:v as CompositeTypeName)").bindparams({
    "v": (field1, field2)
}))
```

ETA2: something like this?
```python
class CustomCast(ColumnElement):
    def __init__(self, left, right):
        self.left = left
        self.right = right

@compiles(CustomCast)
def compile_custom_operator(element, compiler, **kw):
    if not isinstance(element.left, ColumnElement):
        left = literal(element.left)
    else:
        left = element.left
    
    return f"cast({compiler.process(left)} as {element.right})"

q = select(func.funcname(CustomCast(func.row(field1, field2), "CompositeTypeName")))
```

**Reply by zzzeek** [reply]
>Would be nice functionality for sqlalchemy to have, just being able to call a function, maybe open an issue

I'm confused by this entire issue, there seems to be general tone of helplessness all around when all the tools/APIs are right here without any restrictions or limitations.

Call a function, [here's the tutorial](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#working-with-sql-functions).    Casting, [here's the tutorial for cast](https://docs.sqlalchemy.org/en/20/tutorial/data_select.html#tutorial-casts).      

The original issue here seems to be some means of routing data and result columns from a PG composite type into a particular type of Python object structure, the source code to [compositetype](https://github.com/kvesteri/sqlalchemy-utils/blob/master/sqlalchemy_utils/types/pg_composite.py) follows [documented extension patterns](https://docs.sqlalchemy.org/en/20/core/custom_types.html#creating-new-types), it has a clear psycopg2 integration point that can lifted and modified with other DBAPIs with minimal effort.

Usually when comments come in on a very old discussion, the new comment is seeking something different from the original, but overall, sqlalchemy-utils compositetype is easily extended to work on other DBs.   if just needing to emit SQL, that's all in the tutorial.

**Reply by zzzeek** [reply, marked as answer]
> In raw SQL you can write a value with a composite type as `CAST(ROW(field1, field2) AS CompositeTypeName)`. In SQLAlchemy, I guess you can do `func.row(*fields)` but how can you cast to an arbitrary `CompositeTypeName`? Does `execute(text("cast(:row as CompositeTypeName)"), {"row": func.row(*fields)})` work?
> 
> ETA: this works, at least in asyncpg, although the arguments have to be "literal" objects:
> 
> ```python
> func.funcname(text("cast(:v as CompositeTypeName)").bindparams({
>     "v": (field1, field2)
> }))
> ```

here is a demo


```py
from sqlalchemy import cast
from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy.types import UserDefinedType


class CompositeType(UserDefinedType):
    cache_ok = True

    def __init__(self, name, cols):
        self.name = name
        self.cols = tuple(cols)  # must be hashable for caching

    def get_col_spec(self, **kw):
        return self.name

    def bind_expression(self, bindvalue):
        return func.ROW(*bindvalue.value)


expr = cast(
    ("fuzzy dice", 42, 1.99),
    CompositeType("MyCompositeType", ["name", "supplier_id", "price"]),
)

print(select(expr))
print(select(expr).compile(compile_kwargs={"literal_binds": True}))
```

output:

```
SELECT CAST(ROW(:ROW_1, :ROW_2, :ROW_3) AS MyCompositeType) AS anon_1
SELECT CAST(ROW('fuzzy dice', 42, 1.99) AS MyCompositeType) AS anon_1
```
